Split long Telegram messages

aserpi opened the pull request:
Adds the option to split Telegram messages that exceeds current API limits (4096 characters for texts, 200 for captions) into two or more messages.

Pull request #1793 started to dealing with the same issue, but it appears to be abandoned.

Closes #1750

dsander commented:
Sorry @aserpi I totally overlooked this PR. Can you clean up the diff a little bit? It looks like most of the changes were caused by moving methods around.

aserpi commented:
There were two `private` modifier and other inaccurancies in both files, so I cleaned them up.
I rewrote the branch history in order to provide a better understanding of the changes: [this ](https://github.com/aserpi/huginn/compare/master...aserpi:refactor-telegram)is the refactoring diff view and [these](https://github.com/aserpi/huginn/compare/refactor-telegram...aserpi:telegram-split-message) are the actual differences.

dsander commented on app/models/agents/telegram_agent.rb:
```diff
@@ -134,23 +125,52 @@ def load_file(url)
+        params[:text] = params[:text][0..4095] if params[:text]
+        send_message field, params
+      elsif field == :text
+        params[:text].scan(/\G(?:\w{4096}|.{1,4096}(?=\b|\z))/m) do |message|
```
I don't understand the expression 100%, it is trying to split up the text by words first and only falling back to breaking words if it has to, right?

Can you add a spec that would fail with a simpler version like `/.{4096}/m`?

aserpi replied:
I hope so, regular expressions have never been my cup of tea.
Anyway, I rewrote the test to verify word and non-word splitting.

knu commented on app/models/agents/telegram_agent.rb:
```diff
@@ -134,23 +125,52 @@ def load_file(url)
-        params[:disable_web_page_preview] = interpolated['disable_web_page_preview'] if interpolated['disable_web_page_preview'].present?
-        params[:parse_mode] = interpolated['parse_mode'] if interpolated['parse_mode'].present?
+    def send_telegram_messages(field, params)
+      if !interpolated['long_message'].present? || interpolated['long_message'] != 'split'
```
I'd prefer `if interpolated['long_message'] == 'truncate'` for readability.

knu commented on app/models/agents/telegram_agent.rb:
```diff
@@ -33,9 +33,10 @@ class TelegramAgent < Agent
+      * `caption`: caption for a media content (0-200 characters)
       * `disable_notification`: send a message silently in a channel
       * `disable_web_page_preview`: disable link previews for links in a text message
+      * `long_message`: split or truncate text messages or captions that exceeds Telegram API limits
```
What about something like below?  I think users would like to know what the default behavior is.
```
set this to `split` (default) or `truncate` to specify how to deal with messages with a long caption or text that exceeds Telegram API limits
```

aserpi replied:
You are right, I will also add a warning about formatting errors.

Currently the default is `truncate` to avoid the involuntary dispatch of multiple messages in abnormal cases, do you think it should be changed to `split`?

knu replied:
I don't have a strong opinion about it, but I guess `truncate` sounds reasonable as default behavior.

dsander replied:
I agree, if we changed the default we would also need to migrate the options of all existing Agents in the database.

knu commented:
Merged, thanks!
